### DISCUSSION TITLE
Leveraging Quarkus build-time metaprogramming capabilities to improve Jackson's serialization performance

### DISCUSSION THREAD

**Discussion by giscus** [quarkusio/quarkus#42631] [Quarkus Blog/Website]
# Leveraging Quarkus build-time metaprogramming capabilities to improve Jackson's serialization performance

Quarkus: Supersonic Subatomic Java

https://quarkus.io/blog/quarkus-metaprogramming/

**Comment by lost22git** [comment]
It would be better to provide a reflect-free introspection for pojo (just like Micronaut did https://docs.micronaut.io/latest/guide/#introspection)

**Reply by mariofusco** [reply]
This is exactly what Jandex does https://smallrye.io/jandex/

**Reply by lost22git** [reply]
I don't think they implement the same functionality

**Reply by franz1981** [reply]
Exactly; they allow to lookup/set fields without any need for reflection, but the jackson decoding/encoding will just perform big switches on the type of entities + field (name) while serializing/deserializing. Which means a comparison for each field name x type. 
It can work well on benchmarks but once you have TONS of entities and fields, the branch predictor will kill performance.
What we do in Quarkus is more similar to what a user will do manually ie writing a plain sequence of operations on how to map things optimizing string creations too (e.g. field names which are known to be ascii can save been encoded as UTF-8 since they can just be copied as a byte[] into the output stream, unescaping them).
I think that there's very little to improve further performance there (and I rarely say it).

**Comment by mensinda** [comment]
Would it be possible to also enable this feature for non REST endpoints via an annotation? We are heavily using Jackson for serialization / deserialization outside of the REST context. It would be great if we could also use this new generator and ObjectMapper for our non-REST use-case. Most of our beans are auto-generated, so adding an annotation for registering it with Quarkus wouldn't be a problem.

Also, this blog only covers the serialization part. Is reflection-free deserialization also implemented / planned?

**Reply by mariofusco** [reply]
Yes to both questions: we are evaluating how to extend this feature for any JSON serialization and I'm already at work on the deserialization part.

**Comment by Sineaggi** [comment]
How does the performance of standard jackson and this compare to jackson with the [blackbird module](https://github.com/FasterXML/jackson-modules-base/tree/2.18/blackbird)?

**Reply by franz1981** [reply]
Quarkus should still be faster but "depends" where the bottleneck is

**Comment by rquinio1A** [comment]
Very nice example on how to leverage build-time processors in a concrete case!

Do you know what Jackson is using reflection for ?
I always thought reflection was only used once for a given type, and then cached in the static ObjectMapper instance.
If not, it might be worth reporting the perf issue to Jackson project.

**Reply by mariofusco** [reply]
I'm not sure I'm understanding this comment, what do you mean with "cached in the static ObjectMapper"? How can you cache the use of reflection?

**Reply by rquinio1A** [reply]
Thinking about it, Jackson still needs to call `type.newInstance()` and `type.setField(value)`.
What gets cached at runtime by Jackson is more or less what Jandex knows at build time: discovering types, fields, method return types, annotations, etc.

**Reply by mariofusco** [reply]
Correct. Once Jackson has discovered which `java.lang.reflect.Method` has to be called to read a given field of an object (and cached it) it still has to use reflection and call `method.invoke(obj);` every single time which is much slower than call directly obj.get() through the serializers generated at build time.

**Comment by mensinda** [comment]
Does / will this also work with records and classes with public fields only (no getters / setters)?

**Reply by mariofusco** [reply]
Yes.

**Comment by devgauravjava** [comment]
can we do the below with metaprogramming or creating an extension where i can dynamically create kafka consumer based on different application properties because the core logic will be same just topics will be different

**Comment by t1** [comment]
Nice and very promising approach. I just wonder: why did you choose Jackson and not JSON-B? The latter is a standard open for alternative implementations. And the performance of the reference implementation YASSON is even worse than Jackson... by far! And there is a TCK to measure your progress.

**Reply by franz1981** [reply]
Jackson has been used because it allows to programmatically define custom serializers/deserializers and is fairly used by users.
If there are explicit requests for JSON-B feel free to open an issue and we could chat there, thanks!
